ad-aws-login.sh

```diff
@@ -79,31 +85,102 @@ function exit_error() {
 function cleanup() {
   (
     rm -f "${TEMP_FILE}" || true
-    kill "$(pgrep -lf "${USER_BROWSER}.app" | grep "${USER_BROWSER}.*--user-data-dir=${PWD}/user_data$" | awk '{print $1;}')"
+    kill "$BROWSER_PID" || true
   ) &>/dev/null
 }
 
-function handle_browser() {
+function read_os() {
+  case "$OSTYPE" in
+    solaris*) echo "SOLARIS";;
+    darwin*)  echo "OSX";; 
+    linux*)   echo "LINUX";;
+    bsd*)     echo "BSD";;
+    msys*)    echo "WIN";;
+    cygwin*)  echo "WIN";;
+  esac
+}
+
+function find_browser_path() {
+  local list_paths="$1"
+  local list_browsers="$2"
+  local path
   local browser
-  while read -r browser; do
-    if [[ -d "/Applications/${browser}.app" ]]; then
-      readonly USER_BROWSER="${browser}"
-      break
-    fi
-  done < <(echo "$BROWSERS")
+
+  while read -r path; do
+    while read -r browser; do
+      if [[ -d "/${path}/${browser}.app" ]]; then
+        echo "${browser} ${path}"
+        break
+      elif [[ -f "${path}/${browser}" ]]; then
+        echo "${browser} ${path}"
+        break
+      fi
+    done < <(echo "$list_browsers")
+  done < <(echo "$list_paths")
   trap cleanup EXIT
+}
+
+function find_browser() {
+  local os=$(read_os)
+  case "$os" in
+    OSX) echo $(find_browser_path "${FIND_PATHS_OSX}" "${BROWSERS_OSX}");;
+    WIN) echo $(find_browser_path "${FIND_PATHS_WIN}" "${BROWSERS_WIN}");;
+    *)   exit_error 2 "Unhandled OS: ${os}";;
+  esac
+}
+
+function open_browser_osx() {
+  local browser="$1"
+  local url="$3"
+  local args="$4"
+
+  open -a "${browser}" -F -n ${url} --args ${args}
+  local pid=$(pgrep -lf "${browser}\.app.*--user-data-dir=${PWD}/user_data$" | grep -v grep | awk '{print $1;}' || true)
 
-  if [[ -z "${USER_BROWSER}" ]]; then
-    exit_error 1 "Cannot find a browser from:\n${BROWSERS}."
+  echo $pid
+}
+
+function open_browser_posix() {
+  local browser="$1"
+  local path="$2"
+  local url="$3"
+  local args="$4"
+
+  "${path}/${browser}" -F -n ${url} --args ${args} &
+  local pid=$!
+
+  echo $pid
+}
+
+function open_browser() {
+  local os=$(read_os)
+
+  read browser browser_path < <(find_browser)
```

Comment:
This is what happens:
browser="Google"
browser_path="Chrome /Applications"

should be instead:
browser="Google Chrome"
browser_path="/Applications"

This will not work since read command assumes spaces for token separators.